Examples: Added `webgpu_multiple_rendertargets_readback`

Spiri0 opened the pull request:
This is a new example that demonstrates the use of multiple render targets in conjunction with the readRenderTargetPixelsAsync. The readRenderTargetPixelsAsync function was extended by an index with PR #28197 in order to be able to read any texture from a multiple renderTarget

Related issue: #28197

Spiri0 committed:
Add new example and update files.json

Spiri0 committed:
Update files.json

I removed the .html at the end of "webgpu_mrt_readback.html",
now:
"webgpu_mrt_readback",

Spiri0 commented:
I'll improve the example a little later. Although it works exactly as it should, an example should also be as efficient as possible in order to convey an efficient programming style. I would also like to add a size selection so that you can see different readback resolutions.

sunag commented:
Seems that readback normals is not working.
![image](https://github.com/mrdoob/three.js/assets/502810/bc4faf02-ed42-46a3-8a9c-ee84b75df3a5)

Spiri0 commented:
Interesting, it worked for me and aardgoose. 

![3 (5)](https://github.com/mrdoob/three.js/assets/69024222/c3e0dcb7-a52a-4210-952b-f8075a26db07)

Do you have an error message? I'll take a look at it right away

Spiri0 commented:
Oh, has the extension of the readRenderTargetPixelsAsync already been taken into account where you are testing it? This doesn't work without #28197

Spiri0 committed:
Update webgpu_mrt_readback.html

better code. Previously, new instances were constantly being created, which was unnecessary and inefficient.

sunag commented on examples/webgpu_mrt_readback.html:
```diff
@@ -0,0 +1,256 @@
+				if(selector == 2){
+					const pixelBuffer0 = await renderer.readRenderTargetPixelsAsync(readbackTarget, 0, 0, width, height, 0);	//zero is optional
+					
+					const pixelBufferTexture0 = new THREE.DataTexture(pixelBuffer0, width, height);
```
Seems that `DataTexture` and `Material` are created in each frame too

Spiri0 replied:
Ah, something else slipped through my fingers. Yes, of course that shouldn't be the case

Spiri0 replied:
Now it looking good. Have you meanwhile seen the readback_Normal with taking #28197 into account?

Spiri0 commented:
I took the opportunity to make the example a little better. I then deactivated the #28197 extension and then the readback_Normal didn't work for me either. So the same result as yours. That will definitely be the cause for you too.

Spiri0 committed:
Update webgpu_mrt_readback.html

Now no new instances are created during the update, only data is updated as it should be

Spiri0 committed:
Update webgpu_mrt_readback.html

Now no new instances are created during the update, only data is updated as it should be

Spiri0 committed:
Update webgpu_mrt_readback.html

Now no new instances are created during the update, only data is updated as it should be

Mugen87 commented:
FYI: The normal readback does not work for me as well. Tested with the latest commit with Chrome 124.0.6367.80 on macOS.

Mugen87 commented on examples/webgpu_mrt_readback.html:
```diff
@@ -0,0 +1,259 @@
+				const width = readbackTarget.width;
+				const height = readbackTarget.height;
+
+				if(selector == 2){
```
The formatting of the code is still not correct. Please make sure ESLint properly lints the code according to the project's code style.

Spiri0 replied:
My apologies, I thought I did that with the rebase. This is still new for me and I admit there is still a lot I don't know about github.
By formatting do you mean 
if( selector == 2 ) { 
instead of 
if(selector == 2){ 
?

Mugen87 replied:
Yes. And ideally use `===` for strict comparison.

Don't worry if you need time to get things going.

Spiri0 replied:
Thanks for your understanding. I can easily solve the formatting issue. About the rebase, is there anyone I can ask how I do it correctly?

Mugen87 replied:
You can use the merge or rebase command. Normally you start by fetching the latest commit from the upstream repository:
```
git fetch upstream
```
You can merge the latest changes on `dev` into your branch:
```
git merge upstream/dev
```
Or perform a rebase:
```
git rebase upstream/dev
```
The are different resources online which explain the difference between merge and rebase. Both have their pros and cons. However, rebase is ideal for situations when you want to move the point where you have branched to a new starting point. This is exactly what you are looking for in this instance.

Spiri0 committed:
Update webgpu_mrt_readback.html

corrected formatting and strict comparison

Spiri0 committed:
Update webgpu_mrt_readback.html

corrected formatting and strict comparison

Spiri0 committed:
Update webgpu_mrt_readback.html

corrected formatting and strict comparison

Spiri0 committed:
Update webgpu_mrt_readback.html

corrected formatting and strict comparison

sunag committed:
Merge branch 'dev' into pr/28211

sunag committed:
reuse texture() node

sunag committed:
revision

sunag committed:
cleanup

sunag committed:
cleanup

sunag committed:
cleanup

sunag committed:
added info

Spiri0 commented:
I won't get back to it until tonight as I'm at work.

sunag commented:
> I won't get back to it until tonight as I'm at work.

I made some revisions, I think it's ok now. Thank you for the example and fixes :)

sunag committed:
rename webgpu_mrt_readback -> webgpu_multiple_rendertargets_readback

sunag committed:
add tags

Spiri0 commented:
Oh, merged? I haven't done the rebase yet.
